File: enve/Services/Vocabulary/StarDictMorphology.swift
import Foundation

enum StarDictMorphology {

    private static let vowelAccentMap: [Character: Character] = [
        "á": "a", "à": "a", "â": "a", "ä": "a", "ã": "a", "å": "a",
        "é": "e", "è": "e", "ê": "e", "ë": "e",
        "í": "i", "ì": "i", "î": "i", "ï": "i",
        "ó": "o", "ò": "o", "ô": "o", "ö": "o", "õ": "o",
        "ú": "u", "ù": "u", "û": "u", "ü": "u",
        "ý": "y", "ÿ": "y",
    ]

    static func normalize(_ word: String) -> String {
        var out = String.UnicodeScalarView()
        out.reserveCapacity(word.unicodeScalars.count)
        for scalar in word.lowercased().unicodeScalars {
            if let mapped = vowelAccentMap[Character(scalar)] {
                out.append(contentsOf: String(mapped).unicodeScalars)
            } else {
                out.append(scalar)
            }
        }
        return String(out)
    }

    static func candidates(for raw: String) -> [String] {
        let w = raw.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        var seen = Set<String>()
        var out: [String] = []

        func add(_ s: String) {
            guard s.count > 1, seen.insert(s).inserted else { return }
            out.append(s)
        }

        add(w)

        if w.hasSuffix("ily"), w.count > 4 { add(String(w.dropLast(3)) + "y") }
        if w.hasSuffix("ly"), w.count > 4 { add(String(w.dropLast(2))) }

        if w.hasSuffix("ves"), w.count > 4 {
            add(String(w.dropLast(3)) + "f")
            add(String(w.dropLast(3)) + "fe")
        }
        if w.hasSuffix("ies"), w.count > 4 { add(String(w.dropLast(3)) + "y") }

        if w.hasSuffix("ing"), w.count > 5 {
            let base = String(w.dropLast(3))
            add(base + "e")
            add(base)
            if base.count >= 2, base.last == base.dropLast().last {
                add(String(base.dropLast()))
            }
        }

        if w.hasSuffix("ied"), w.count > 4 { add(String(w.dropLast(3)) + "y") }
        if w.hasSuffix("ed"), w.count > 3 {
            let base = String(w.dropLast(2))
            add(String(w.dropLast()))
            add(base)
            if base.count >= 2, base.last == base.dropLast().last {
                add(String(base.dropLast()))
            }
        }

        if w.hasSuffix("est"), w.count > 5 {
            let base = String(w.dropLast(3))
            add(base + "e")
            add(base)
            if base.count >= 2, base.last == base.dropLast().last {
                add(String(base.dropLast()))
            }
        }
        if w.hasSuffix("er"), w.count > 4 {
            let base = String(w.dropLast(2))
            add(base + "e")
            add(base)
            if base.count >= 2, base.last == base.dropLast().last {
                add(String(base.dropLast()))
            }
        }

        if w.hasSuffix("es"), w.count > 3 {
            add(String(w.dropLast()))
            add(String(w.dropLast(2)))
        }
        if w.hasSuffix("s"), w.count > 3 {
            add(String(w.dropLast()))
        }

        if w.count >= 3 {
            add(w + "a")
            add(w + "e")
            add(w + "i")

            if w.hasSuffix("o"), w.count > 3 {
                add(String(w.dropLast()))
                add(String(w.dropLast()) + "a")
            }

            if w.hasSuffix("i"), w.count > 3 {
                add(String(w.dropLast()) + "a")
                add(String(w.dropLast()) + "e")
            }
            if w.hasSuffix("u"), w.count > 3 {
                add(String(w.dropLast()))
                add(String(w.dropLast()) + "a")
                add(String(w.dropLast()) + "e")
            }

            if w.hasSuffix("ji"), w.count > 3 { add(String(w.dropLast(2))) }
            if w.hasSuffix("evi"), w.count > 4 {
                add(String(w.dropLast(3)))
                add(String(w.dropLast(3)) + "e")
            }
            if w.hasSuffix("ovi"), w.count > 4 { add(String(w.dropLast(3))) }
            if w.hasSuffix("je"), w.count > 3 { add(String(w.dropLast(2))) }
            if w.hasSuffix("i"), w.count > 3 { add(String(w.dropLast())) }
            if w.hasSuffix("e"), w.count > 3 { add(String(w.dropLast()) + "a") }

            if w.hasSuffix("ega"), w.count > 4 { add(String(w.dropLast(3)) + "i") }
            if w.hasSuffix("emu"), w.count > 4 { add(String(w.dropLast(3)) + "i") }
            if w.hasSuffix("em"), w.count > 3 { add(String(w.dropLast(2)) + "i") }
            if w.hasSuffix("ih"), w.count > 3 { add(String(w.dropLast(2)) + "i") }

            if w.hasSuffix("ati"), w.count > 4 { add(String(w.dropLast(3))) }
            if w.hasSuffix("iti"), w.count > 4 { add(String(w.dropLast(3))) }
            if w.hasSuffix("eti"), w.count > 4 { add(String(w.dropLast(3))) }

            if w.hasSuffix("am"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s); add(s + "ti"); add(s + "ati")
            }
            if w.hasSuffix("aš"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s); add(s + "ti"); add(s + "ati")
            }
            if w.hasSuffix("iš"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s); add(s + "ti"); add(s + "iti")
            }
            if w.hasSuffix("im"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s); add(s + "ti"); add(s + "iti")
            }
            if w.hasSuffix("eš"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s); add(s + "ti"); add(s + "eti")
            }
            if w.hasSuffix("em"), w.count > 3 {
                let s = String(w.dropLast(2))
                add(s + "eti")
            }
        }

        return out
    }
}

File: android/engine/src/main/java/com/enve/app/data/vocab/StarDictMorphology.kt
package com.enve.app.data.vocab

object StarDictMorphology {

    private val vowelAccentMap: Map<Char, Char> = mapOf(
        'á' to 'a', 'à' to 'a', 'â' to 'a', 'ä' to 'a', 'ã' to 'a', 'å' to 'a',
        'é' to 'e', 'è' to 'e', 'ê' to 'e', 'ë' to 'e',
        'í' to 'i', 'ì' to 'i', 'î' to 'i', 'ï' to 'i',
        'ó' to 'o', 'ò' to 'o', 'ô' to 'o', 'ö' to 'o', 'õ' to 'o',
        'ú' to 'u', 'ù' to 'u', 'û' to 'u', 'ü' to 'u',
        'ý' to 'y', 'ÿ' to 'y',
    )

    fun normalize(word: String): String = buildString(word.length) {
        for (c in word.lowercase()) {
            append(vowelAccentMap[c] ?: c)
        }
    }

    fun candidates(raw: String): List<String> {
        val w = raw.lowercase().trim()
        val seen = HashSet<String>()
        val out = mutableListOf<String>()

        fun add(s: String) {
            if (s.length > 1 && seen.add(s)) out.add(s)
        }

        add(w)

        if (w.endsWith("ily") && w.length > 4) add(w.dropLast(3) + "y")
        if (w.endsWith("ly") && w.length > 4) add(w.dropLast(2))

        if (w.endsWith("ves") && w.length > 4) {
            add(w.dropLast(3) + "f")
            add(w.dropLast(3) + "fe")
        }
        if (w.endsWith("ies") && w.length > 4) add(w.dropLast(3) + "y")

        if (w.endsWith("ing") && w.length > 5) {
            val base = w.dropLast(3)
            add(base + "e")
            add(base)
            if (base.length >= 2 && base.last() == base[base.length - 2]) {
                add(base.dropLast(1))
            }
        }

        if (w.endsWith("ied") && w.length > 4) add(w.dropLast(3) + "y")
        if (w.endsWith("ed") && w.length > 3) {
            val base = w.dropLast(2)
            add(w.dropLast(1))
            add(base)
            if (base.length >= 2 && base.last() == base[base.length - 2]) {
                add(base.dropLast(1))
            }
        }

        if (w.endsWith("est") && w.length > 5) {
            val base = w.dropLast(3)
            add(base + "e")
            add(base)
            if (base.length >= 2 && base.last() == base[base.length - 2]) {
                add(base.dropLast(1))
            }
        }
        if (w.endsWith("er") && w.length > 4) {
            val base = w.dropLast(2)
            add(base + "e")
            add(base)
            if (base.length >= 2 && base.last() == base[base.length - 2]) {
                add(base.dropLast(1))
            }
        }

        if (w.endsWith("es") && w.length > 3) {
            add(w.dropLast(1))
            add(w.dropLast(2))
        }
        if (w.endsWith("s") && w.length > 3) add(w.dropLast(1))

        if (w.length >= 3) {
            add(w + "a")
            add(w + "e")
            add(w + "i")

            if (w.endsWith("o") && w.length > 3) {
                add(w.dropLast(1))
                add(w.dropLast(1) + "a")
            }

            if (w.endsWith("i") && w.length > 3) {
                add(w.dropLast(1) + "a")
                add(w.dropLast(1) + "e")
            }
            if (w.endsWith("u") && w.length > 3) {
                add(w.dropLast(1))
                add(w.dropLast(1) + "a")
                add(w.dropLast(1) + "e")
            }

            if (w.endsWith("ji") && w.length > 3) add(w.dropLast(2))
            if (w.endsWith("evi") && w.length > 4) {
                add(w.dropLast(3))
                add(w.dropLast(3) + "e")
            }
            if (w.endsWith("ovi") && w.length > 4) add(w.dropLast(3))
            if (w.endsWith("je") && w.length > 3) add(w.dropLast(2))
            if (w.endsWith("i") && w.length > 3) add(w.dropLast(1))
            if (w.endsWith("e") && w.length > 3) add(w.dropLast(1) + "a")

            if (w.endsWith("ega") && w.length > 4) add(w.dropLast(3) + "i")
            if (w.endsWith("emu") && w.length > 4) add(w.dropLast(3) + "i")
            if (w.endsWith("em") && w.length > 3) add(w.dropLast(2) + "i")
            if (w.endsWith("ih") && w.length > 3) add(w.dropLast(2) + "i")

            if (w.endsWith("ati") && w.length > 4) add(w.dropLast(3))
            if (w.endsWith("iti") && w.length > 4) add(w.dropLast(3))
            if (w.endsWith("eti") && w.length > 4) add(w.dropLast(3))

            if (w.endsWith("am") && w.length > 3) {
                val s = w.dropLast(2); add(s); add(s + "ti"); add(s + "ati")
            }
            if (w.endsWith("aš") && w.length > 3) {
                val s = w.dropLast(2); add(s); add(s + "ti"); add(s + "ati")
            }
            if (w.endsWith("iš") && w.length > 3) {
                val s = w.dropLast(2); add(s); add(s + "ti"); add(s + "iti")
            }
            if (w.endsWith("im") && w.length > 3) {
                val s = w.dropLast(2); add(s); add(s + "ti"); add(s + "iti")
            }
            if (w.endsWith("eš") && w.length > 3) {
                val s = w.dropLast(2); add(s); add(s + "ti"); add(s + "eti")
            }
            if (w.endsWith("em") && w.length > 3) {
                val s = w.dropLast(2); add(s + "eti")
            }
        }

        return out
    }
}
